elsa-types.el

```diff
@@ -476,5 +476,125 @@ other, then this is a supertype of other."
 (cl-defmethod elsa-type-describe ((this elsa-generic-type))
   (symbol-name (oref this label)))
 
+;; One-dimensional sparse arrays indexed by characters
+(defclass elsa-type-charttable (elsa-type) ())
+
+(cl-defmethod elsa-type-describe ((this elsa-type-charttable))
+  "Charttable")
+
+;; One-dimensional arrays of t or nil.
+(defclass elsa-type-boolvector (elsa-type) ())
+
+(cl-defmethod elsa-type-describe ((this elsa-type-boolvector))
+  "Boolvector")
+
+;; Super-fast lookup tables
+(defclass elsa-type-hashtable (elsa-type) ())
+
+(cl-defmethod elsa-type-describe ((this elsa-type-hashtable))
+  "Hashtable")
+
+;; A piece of executable code you can call from elsewhere.
+(defclass elsa-type-function (elsa-type) ())
+
+(cl-defmethod elsa-type-describe ((this elsa-type-function))
+  "Function")
+
+;; A method of expanding an expression into another expression, more fundamental but less pretty
+(defclass elsa-type-macro (elsa-type) ())
+
+(cl-defmethod elsa-type-describe ((this elsa-type-macro))
+  "Macro")
+
+;; A function written in C, callable from Lisp
+(defclass elsa-type-primitive (elsa-type) ())
```

Comment:
So there is `subrp` which can decide on these.  So it makes sense to be able to distinguish it but maybe we should extend the function type then.

Reply:
I removed it from my commit. We can improve the function type later.